### PR TITLE
Travel guides

### PR DESCRIPTION
**Pull request by vshcherb**

### CODE REVIEW

File: OsmAnd.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -400,6 +400,7 @@
 		326B770A29D9CA6400FB23DF /* OATripRecordingTimeWidget.mm in Sources */ = {isa = PBXBuildFile; fileRef = 326B770929D9CA6400FB23DF /* OATripRecordingTimeWidget.mm */; };
```
Comment (nnngrach) [inline OsmAnd.xcodeproj/project.pbxproj:400]: Screenshot 2023-08-25 at 19 43 30

File: Sources/Controllers/TravelGuides/article/TravelGuidesContentsViewController.swift [hunk 2]
```diff
@@ -13,18 +13,26 @@ class TravelGuidesContentsViewController : OABaseButtonsViewController {
     var article: TravelArticle
```
Comment (nnngrach) [inline Sources/Controllers/TravelGuides/article/TravelGuidesContentsViewController.swift:13]: Screenshot 2023-08-25 at 21 55 55
Screenshot 2023-08-25 at 21 56 03

File: Sources/Controllers/TravelGuides/settings/TravelGuidesSettingsViewController.swift [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        downloadImagesRow.cellType = OAValueTableViewCell.getIdentifier()
+        downloadImagesRow.title = localizedString("wikivoyage_download_pics")
+        downloadImagesRow.descr = OsmAndApp.swiftInstance().data.travelGuidesImagesDownloadMode.title
+        downloadImagesRow.iconName = "ic_custom_photo"
+        downloadImagesRow.key = "downloadImagesRow"
+        
+        let cacheSizeRow = imagesSection.createNewRow()
+        cacheSizeRow.cellType = OAValueTableViewCell.getIdentifier()
+        cacheSizeRow.title = localizedString("cache_size")
+        cacheSizeRow.descr = "123 MB"
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/settings/TravelGuidesSettingsViewController.swift:41]: ?

File: Sources/Controllers/TravelGuides/TravelGuidesUtils.swift [hunk 4]
```diff
@@ -8,20 +8,22 @@
...
     
     static func processWikivoyageDomain(url: String, delegate: TravelArticleDialogProtocol) {
-        let lang = OAWikiArticleHelper.getLang(url)
-        let articleName = OAWikiArticleHelper.getArticleName(fromUrl: url, lang: lang)
-        let articleId = TravelObfHelper.shared.getArticleId(title: articleName!, lang: lang!)
-        if articleId != nil {
-            delegate.openArticleById(articleId: articleId!, selectedLang: lang!)
-        } else {
-            OAWikiArticleHelper.warnAboutExternalLoad(url, sourceView: delegate.getWebView())
+        if let lang = OAWikiArticleHelper.getLang(url) {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/TravelGuidesUtils.swift:18]: guard let lang = OAWikiArticleHelper.getLang(url),
let articleName = OAWikiArticleHelper.getArticleName(fromUrl: url, lang: lang) else {
    return
}

File: Sources/Controllers/TravelGuides/TravelGuidesUtils.swift [hunk 5]
```diff
@@ -49,12 +53,13 @@ class TravelGuidesUtils {
             if fristValueStart != -1 && firstValueEnd != -1 && secondValueStart != -1  && firstValueEnd > fristValueStart {
                 let lat = geoPart.substring(from: fristValueStart + 1, to: firstValueEnd)
                 let lon = secondGeoPart.substring(from: fristValueStart)
-                if Double(lat) != nil && Double(lon) != nil {
-                    return CLLocation(latitude: Double(lat)!, longitude: Double(lon)!)
+                if let doubleLat = Double(lat) {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/TravelGuidesUtils.swift:56]: if let doubleLat = Double(lat), let doubleLon = Double(lon) {}

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 6]
```diff
@@ -146,71 +146,78 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
     
     func setupBottomButtonsView() {
         bottomView = UIView()
-        bottomView!.addBlurEffect(true, cornerRadius: 0, padding: 0)
-        view.addSubview(bottomView!)
+        guard let bottomView else {return}
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:149]: { return }

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 7]
```diff
@@ -146,71 +146,78 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
...
+        bottomView.addBlurEffect(true, cornerRadius: 0, padding: 0)
+        view.addSubview(bottomView)
         
         bottomStackView = UIStackView()
-        bottomStackView!.axis = .horizontal
-        bottomStackView!.alignment = .center
-        bottomStackView!.distribution = .equalCentering
-        bottomStackView!.spacing = 8
-        bottomView!.addSubview(bottomStackView!)
+        guard let bottomStackView else {return}
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:154]: { return }

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 8]
```diff
@@ -146,71 +146,78 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
...
+        bottomStackView.spacing = 8
+        bottomView.addSubview(bottomStackView)
         
         contentButton = UIButton()
-        contentButton!.setImage(UIImage.templateImageNamed("ic_custom_list"), for: .normal)
-        contentButton!.tintColor = UIColor.iconColorActive
-        contentButton!.contentHorizontalAlignment = .left
-        contentButton!.addTarget(self, action: #selector(self.onContentsButtonClicked), for: .touchUpInside)
-        bottomStackView!.addArrangedSubview(contentButton!)
+        guard let contentButton else {return}
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:162]: { return }

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 9]
```diff
@@ -146,71 +146,78 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
...
+        bottomStackView.addArrangedSubview(bookmarkButton)
         updateBookmarkButton()
     }
     
     func updateBookmarkButton() {
         if let article {
-            let isSaved = TravelObfHelper.shared.getBookmarksHelper().isArticleSaved(article: article)
-            let iconName = isSaved ? "ic_navbar_bookmark" : "ic_navbar_bookmark_outlined"
-            bookmarkButton!.setImage(UIImage.templateImageNamed(iconName), for: .normal)
+            if let bookmarkButton {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:193]: If the check is done at the beginning of the function, it's better to use the 'guard let' construct

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 10]
```diff
@@ -146,71 +146,78 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
...
+                let isSaved = TravelObfHelper.shared.getBookmarksHelper().isArticleSaved(article: article)
+                let iconName = isSaved ? "ic_navbar_bookmark" : "ic_navbar_bookmark_outlined"
+                bookmarkButton.setImage(UIImage(named:iconName), for: .normal)
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:196]: code style UIImage(named: iconName)

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 11]
```diff
@@ -456,19 +462,20 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
...
         
         
-        if imageTitle != nil && imageTitle!.length > 0 {
-            let imagesDownloadMode = getImagesDownloadMode()
-            let dontLoadImages = !self.isDownloadImagesOnlyNow() && (imagesDownloadMode!.isDontDownload() || (imagesDownloadMode!.isDownloadOnlyViaWifi() && AFNetworkReachabilityManagerWrapper.isReachableViaWWAN()))
-            
-            if !dontLoadImages {
-                let url = TravelArticle.getImageUrl(imageTitle: imageTitle!, thumbnail: false)
-                sb += "<div class=\"title-image" + nightModeClass + "\" style=\"background-image: url(" + url + ")\"></div>"
+        if imageTitle != nil && !imageTitle!.isEmpty {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:465]: force unwrap

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 12]
```diff
@@ -477,45 +484,47 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
...
+                {
+                    self.bottomStackView!.addSpinner(inCenterOfCurrentView: true)
                     self.pointsButton!.setTitle("", for: .normal)
+                    self.pointsButton!.setImage(nil, for: .normal)
                     self.pointsButton!.isEnabled = false
                 }
-                self.bottomStackView!.removeSpinner()
+                else
+                {
+                    if gpxFile != nil && gpxFile!.pointsCount() > 0 {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:513]: force unwrap

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 13]
```diff
@@ -560,25 +567,35 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
...
+        if let url = URL(string: link) {
+            webView.load(URLRequest(url: url))
+        }
     }
     
     func openArticleByTitle(title: String, selectedLang: String) {
-        historyArticleIds.append(self.articleId!)
-        historyLangs.append(self.selectedLang!)
-        self.articleId = TravelObfHelper.shared.getArticleId(title: title, lang: selectedLang)
+        if let currentArticleId = self.articleId {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:581]: remove self

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 14]
```diff
@@ -560,25 +567,35 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
...
+        }
+        articleId = TravelObfHelper.shared.getArticleId(title: title, lang: selectedLang)
         self.selectedLang = selectedLang
         populateArticle()
     }
     
     func openArticleById(articleId: TravelArticleIdentifier, selectedLang: String) {
-        historyArticleIds.append(self.articleId!)
-        historyLangs.append(self.selectedLang!)
+        if let currentArticleId = self.articleId {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:593]: remove self

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 15]
```diff
@@ -560,25 +567,35 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
...
+            historyArticleIds.append(currentArticleId)
+        }
+        if let currentSelectedLang = self.selectedLang {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:596]: self

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 16]
```diff
@@ -588,9 +605,13 @@ class TravelArticleDialogViewController : OABaseWebViewController, TravelArticle
     //MARK: OAWikiLanguagesWebDelegate
     
     func onLocaleSelected(_ locale: String!) {
-        historyArticleIds.append(self.articleId!)
-        historyLangs.append(self.selectedLang!)
-        self.selectedLang = locale
+        if let currentArticleId = self.articleId {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:608]: self

File: Sources/Controllers/TravelGuides/article/TravelGuidesContentsViewController.swift [hunk 17]
```diff
@@ -29,31 +29,28 @@ class TravelGuidesContentsViewController : OABaseNavbarViewController {
     override func generateData() {
         tableData.clearAllData()
         
-        var displayingItems = items!.subItems
-        if selectedSubitemIndex != nil {
-            displayingItems = items!.subItems[selectedSubitemIndex!].subItems
-        }
+        guard let items else {return}
+        var displayingItems = selectedSubitemIndex != nil ? items.subItems[selectedSubitemIndex!].subItems : items.subItems
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelGuidesContentsViewController.swift:33]: force unwrap selectedSubitemIndex

File: Sources/Controllers/TravelGuides/article/TravelGuidesContentsViewController.swift [hunk 18]
```diff
@@ -29,31 +29,28 @@ class TravelGuidesContentsViewController : OABaseNavbarViewController {
...
-            } else {
-                row.iconName = "ic_custom_book_info"
-            }
-            
-            if item.subItems.count > 0 {
-                row.setObj(true, forKey: "hasSubitems")
+            if let itemLink = item.link {
+                row.setObj(itemLink, forKey: "link")
+                
+                if item.parent != nil && item.parent!.link != nil {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelGuidesContentsViewController.swift:43]: force unwrap

File: Sources/Controllers/TravelGuides/article/TravelGuidesContentsViewController.swift [hunk 19]
```diff
@@ -121,7 +117,7 @@ class TravelGuidesContentsViewController : OABaseNavbarViewController {
                 
                 let hasSubitems = item.bool(forKey: "hasSubitems")
                 if hasSubitems {
-                    cell.button.setImage(UIImage.templateImageNamed("ic_custom_arrow_right"), for: .normal)
+                    cell.button.setImage(UIImage(named:"ic_custom_arrow_right"), for: .normal)
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelGuidesContentsViewController.swift:120]: UIImage(named: "ic_custom_arrow_right")

File: Sources/Controllers/TravelGuides/article/TravelGuidesNavigationViewController.swift [hunk 20]
```diff
@@ -57,69 +58,68 @@ class TravelGuidesNavigationViewController : OABaseNavbarViewController {
...
-        if let parts = article!.aggregatedPartOf {
-            names = parts
-                .split(separator: ",")
-                .reversed()
-                .map({ substring in
-                    return String(substring)
-                })
-            if navigationMap.count > names.count && article!.title != nil {
-                names.append(article!.title!)
+        if let article {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelGuidesNavigationViewController.swift:103]: if let article, let parts = article.aggregatedPartOf

File: Sources/Controllers/TravelGuides/article/TravelGuidesNavigationViewController.swift [hunk 21]
```diff
@@ -57,69 +58,68 @@ class TravelGuidesNavigationViewController : OABaseNavbarViewController {
...
+            if let parts = article.aggregatedPartOf {
+                names = parts
+                    .split(separator: ",")
+                    .reversed()
+                    .map({ substring in
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelGuidesNavigationViewController.swift:108]: .map { String($0) }

File: Sources/Controllers/TravelGuides/article/TravelGuidesNavigationViewController.swift [hunk 22]
```diff
@@ -207,35 +209,40 @@ class TravelGuidesNavigationViewController : OABaseNavbarViewController {
...
+                        let vc = TravelGuidesNavigationViewController()
+                        vc.setupWith(article: article, selectedLang: selectedLang, navigationMap: navigationMap, regionsNames: [], selectedItem: item)
+                        vc.delegate = delegate
+                        navigationController?.pushViewController(vc, animated: true)
+                    }
                 }
             } else {
-                if delegate != nil {
-                    delegate!.openArticleByTitle(title: articleId.title!, selectedLang: selectedLang)
+                if let delegate {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelGuidesNavigationViewController.swift:220]: if let delegate, let title = articleId.title

File: Sources/Controllers/TravelGuides/data/TravelObfHelper.swift [hunk 23]
```diff
@@ -135,8 +139,10 @@ class TravelObfHelper : NSObject {
         
         var results: [OAFoundAmenity] = []
         func publish(poi: OAPOIAdapter?) -> Bool {
-            if lang == nil || lang?.length == 0 || poi!.getNamesMap(true)!.keys.contains(lang!) {
-                results.append(OAFoundAmenity(file: reader, amenity: poi))
+            if let poi {
+                if lang == nil || lang!.length == 0 || (poi.getNamesMap(true)?.keys.contains(lang!) ?? false) {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/data/TravelObfHelper.swift:143]: force unwrap

File: Sources/Controllers/TravelGuides/data/TravelObfHelper.swift [hunk 24]
```diff
@@ -148,16 +154,21 @@ class TravelObfHelper : NSObject {
...
-            articles = readArticles(file: file!, amenity: amenity)
-        }
-        if articles != nil && articles!.count > 0 {
-            var i = articles!.values.makeIterator()
-            let newArticleId = i.next()!.generateIdentifier()
-            cachedArticles[newArticleId.hashValue] = articles
-            article = getCachedArticle(articleId: newArticleId, lang: lang, readGpx: readPoints, callback: callback)
+            articles = readArticles(file: file, amenity: amenity)
+        }
+        if let articles {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/data/TravelObfHelper.swift:163]: if let articles, !articles.isEmpty

File: Sources/Controllers/TravelGuides/data/TravelObfHelper.swift [hunk 25]
```diff
@@ -381,33 +392,35 @@ class TravelObfHelper : NSObject {
         var navMap = [String : [TravelSearchResult]]()
         var headers = [String]()
         var headerObjs = [String : TravelSearchResult]()
-        if parts.count > 0 {
+        if !parts.isEmpty {
             headers.append(contentsOf: parts)
-            if article.isParentOf != nil && !article.isParentOf!.isEmpty {
-                headers.append(title!)
+            if let isParentOf = article.isParentOf {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/data/TravelObfHelper.swift:397]: if let isParentOf = article.isParentOf, !isParentOf.isEmpty

File: Sources/Controllers/TravelGuides/data/TravelObfHelper.swift [hunk 26]
```diff
@@ -434,14 +447,16 @@ class TravelObfHelper : NSObject {
         
         for reader in getReaders() {
             OATravelGuidesHelper.searchAmenity(title, categoryName: ROUTE_ARTICLE, radius: -1, lat: -1, lon: -1, reader: reader) { amenity in
-                if title == amenity!.getName(lang, transliterate: false) {
-                    amenities.append(amenity!)
-                    return true
+                if let amenity {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/data/TravelObfHelper.swift:450]: you need to make one if

File: Sources/Controllers/TravelGuides/data/TravelObfHelper.swift [hunk 27]
```diff
@@ -711,14 +747,16 @@ class TravelObfHelper : NSObject {
             
             OATravelGuidesHelper.searchAmenity(x, y: y, left: left, right: right, top: top, bottom: bottom, reader: reader, searchFilter: ROUTE_ARTICLE) { amenity in
                 
-                if title == amenity!.getName(lang, transliterate: false) {
-                    amenities.append(amenity!)
-                    return true
+                if let amenity {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/data/TravelObfHelper.swift:750]: you need to make one if

File: Sources/Controllers/TravelGuides/data/TravelObfHelper.swift [hunk 28]
```diff
@@ -831,20 +867,27 @@ class GpxFileReader {
...
-        if callback != nil {
-            callback!.onGpxFileReading()
+        if let callback {
+            callback.onGpxFileReading?()
         }
     }
     
     func doInBackground() -> OAGPXDocumentAdapter? {
-        return TravelObfHelper.shared.buildGpxFile(readers: readers!, article: article!)
+        if let readers {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/data/TravelObfHelper.swift:876]: 1 if

File: Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift [hunk 29]
```diff
@@ -181,25 +189,26 @@ class TravelExploreViewController: OABaseNavbarViewController, TravelExploreView
...
-                            gpxRow.title = title
-                            gpxRow.descr = item.user
-                            gpxRow.setObj(item, forKey: "article")
-                            
-                            let analysis = item.getAnalysis()
-                            let statisticsCells = OATrackMenuHeaderView.generateGpxBlockStatistics(analysis, withoutGaps: false)
-                            gpxRow.setObj(statisticsCells, forKey: "statistics_cells")
+                            if let item: TravelGpx = article as? TravelGpx {
+                                let gpxRow = articlesSection.createNewRow()
+                                let title = (item.descr != nil && !item.descr!.isEmpty) ? item.descr! : item.title
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift:201]: unwrap optional

File: Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift [hunk 30]
```diff
@@ -211,7 +220,7 @@ class TravelExploreViewController: OABaseNavbarViewController, TravelExploreView
                             articleRow.setObj(item.getGeoDescription() ?? "", forKey: "isPartOf")
                             articleRow.setObj(item, forKey: "article")
                             articleRow.setObj(item.lang, forKey: "lang")
-                            if (item.imageTitle != nil && item.imageTitle!.length > 0) {
+                            if item.imageTitle != nil && item.imageTitle!.length > 0 {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift:223]: unwrap optional

File: Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift [hunk 31]
```diff
@@ -282,7 +291,7 @@ class TravelExploreViewController: OABaseNavbarViewController, TravelExploreView
                     resultRow.setObj(item.articleId, forKey: "articleId")
                     
                     resultRow.setObj("ic_custom_photo", forKey: "noImageIcon")
-                    if (item.imageTitle != nil && item.imageTitle!.length > 0) {
+                    if item.imageTitle != nil && item.imageTitle!.length > 0 {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift:294]: unwrap optional

File: Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift [hunk 32]
```diff
@@ -295,39 +304,40 @@ class TravelExploreViewController: OABaseNavbarViewController, TravelExploreView
...
         downloadingCellHelper.hostDataLock = dataLock
-        weak var weakself = self
         
-        downloadingCellHelper.fetchResourcesBlock = {
+        downloadingCellHelper.fetchResourcesBlock = { [weak self] in
+            guard let self else { return }
             var downloadingResouces = OAResourcesUISwiftHelper.getResourcesInRepositoryIds(byRegionId: "travel", resourceTypeNames: ["travel"])
-            if (downloadingResouces != nil) {
+            if downloadingResouces != nil {
                 downloadingResouces!.sort(by: { a, b in
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift:312]: unwrap optional

File: Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift [hunk 33]
```diff
@@ -295,39 +304,40 @@ class TravelExploreViewController: OABaseNavbarViewController, TravelExploreView
...
         
-        downloadingCellHelper.getSwiftResourceByIndexBlock = { (indexPath: IndexPath?) -> OAResourceSwiftItem? in
-            
-            let headerCellsCountInResourcesSection = weakself!.headerCellsCountInResourcesSection()
-            if (indexPath != nil && indexPath!.row >= headerCellsCountInResourcesSection) {
-                return weakself!.downloadingResources[indexPath!.row - headerCellsCountInResourcesSection]
+        downloadingCellHelper.getSwiftResourceByIndexBlock = { [weak self] (indexPath: IndexPath?) -> OAResourceSwiftItem? in
+            guard let self else { return nil }
+            let headerCellsCountInResourcesSection = self.headerCellsCountInResourcesSection()
+            if indexPath != nil && indexPath!.row >= headerCellsCountInResourcesSection {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift:322]: unwrap optional

File: Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift [hunk 34]
```diff
@@ -486,48 +497,52 @@ class TravelExploreViewController: OABaseNavbarViewController, TravelExploreView
...
+                cell.article = item.obj(forKey: "article") as? TravelArticle
+                cell.articleLang = item.string(forKey: "lang")
+                
+                cell.arcticleTitle.text = item.title
+                cell.arcticleDescription.text = item.descr
+                cell.regionLabel.text = item.string(forKey: "isPartOf")
+                
+                cell.imageVisibility(true)
+                if item.iconName != nil && !notDownloadImages() {
+                    startAsyncImageDownloading(item.iconName!, cell)
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/explore/TravelExploreViewController.swift:513]: unwrap optional

File: Sources/Controllers/TravelGuides/TravelGuidesUtils.swift [hunk 35]
```diff
@@ -15,18 +15,18 @@ final class TravelGuidesUtils {
...
-                    delegate.openArticleById(articleId: articleId, selectedLang: lang)
-                } else {
-                    OAWikiArticleHelper.warnAboutExternalLoad(url, sourceView: delegate.getWebView())
-                }
-            }
+        guard let lang = OAWikiArticleHelper.getLang(url), 
+                let articleName = OAWikiArticleHelper.getArticleName(fromUrl: url, lang: lang) else { return }
+        
+        let articleId = TravelObfHelper.shared.getArticleId(title: articleName, lang: lang)
+        if let articleId {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/TravelGuidesUtils.swift:22]: if let articleId = TravelObfHelper.shared.getArticleId(title: articleName, lang: lang) {}

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 36]
```diff
@@ -469,13 +464,15 @@ final class TravelArticleDialogViewController : OABaseWebViewController, TravelA
...
         
         
-        if imageTitle != nil && !imageTitle!.isEmpty {
-            if let imagesDownloadMode = getImagesDownloadMode() {
-                let dontLoadImages = !self.isDownloadImagesOnlyNow() && (imagesDownloadMode.isDontDownload() || (imagesDownloadMode.isDownloadOnlyViaWifi() && AFNetworkReachabilityManagerWrapper.isReachableViaWWAN()))
-                
-                if !dontLoadImages {
-                    let url = TravelArticle.getImageUrl(imageTitle: imageTitle!, thumbnail: false)
-                    sb += "<div class=\"title-image" + nightModeClass + "\" style=\"background-image: url(" + url + ")\"></div>"
+        if let imageTitle {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:467]: if let imageTitle, !imageTitle.isEmpty, let imagesDownloadMode = getImagesDownloadMode() {}

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 37]
```diff
@@ -517,13 +514,14 @@ final class TravelArticleDialogViewController : OABaseWebViewController, TravelA
...
                 else
                 {
-                    if gpxFile != nil && gpxFile!.pointsCount() > 0 {
-                        let title = localizedString("shared_string_gpx_points") + ": " + String(gpxFile!.pointsCount())
-                        self.pointsButton!.setTitle(title , for: .normal)
-                        self.pointsButton!.isEnabled = true
-                    } else {
-                        self.pointsButton!.setTitle("", for: .normal)
-                        self.pointsButton!.isEnabled = false
+                    self.pointsButton!.setTitle("", for: .normal)
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:517]: force

File: Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift [hunk 38]
```diff
@@ -517,13 +514,14 @@ final class TravelArticleDialogViewController : OABaseWebViewController, TravelA
...
+                    self.pointsButton!.isEnabled = false
+                    if let gpxFile {
```
Comment (tigrim) [inline Sources/Controllers/TravelGuides/article/TravelArticleDialogViewController.swift:519]: let gpxFile, pxFile.pointsCount() > 0

**Review by tigrim — APPROVED** [review]

File: Sources/Controllers/TravelGuides/data/OATravelLocalDataDbHelper.mm [hunk 39]
```diff
@@ -663,9 +644,21 @@ - (void) updateSavedArticle:(OATravelArticle *)oldArticle newArticle:(OATravelAr
...
 
+- (sqlite3 *) getDbInstance
+{
+    return _dbInstance;
+}
+
+- (dispatch_queue_t) getDbQueue
+{
+    return _dbQueue;
+}
```
Comment (alex-osm) [inline Sources/Controllers/TravelGuides/data/OATravelLocalDataDbHelper.mm:662]: Redundant?

File: Sources/Helpers/OAMapAlgorithms.mm [hunk 40]
```diff
@@ -0,0 +1,101 @@
...
+        else
+        {
+            res.push_back(-(num >> 1));
+        }
+    }
+    
+    return res;
+}
+
++ (OATrkSegment *) augmentTrkSegmentWithAltitudes:(OATrkSegment *)sgm decodedSteps:(QList<int>)decodedSteps startEle:(double)startEle
```
Comment (alex-osm) [inline Sources/Helpers/OAMapAlgorithms.mm:45]: Fix

File: Sources/Helpers/OAWebImagesCacheHelper.mm [hunk 41]
```diff
@@ -0,0 +1,322 @@
...
+- (NSString *) getDbFilename
+{
+    return nil;
+}
+
+// Overriding in sublcass
+- (NSString *) getDbFoldername
+{
+    return nil;
+}
```
Comment (alex-osm) [inline Sources/Helpers/OAWebImagesCacheHelper.mm:64]: Redundant?

File: Sources/GPX/OAGPXDocument.mm [hunk 42]
```diff
@@ -616,7 +616,7 @@ + (void)fillRoute:(std::shared_ptr<OsmAnd::GpxDocument::Route>)rte usingRoute:(O
     [r fillExtensions:rte];
 }
 
-- (BOOL) saveTo:(NSString *)filename
+- (std::shared_ptr<OsmAnd::GpxDocument> ) getPrepareForSavingDocument
```
Comment (alex-osm) [inline Sources/GPX/OAGPXDocument.mm:619]: Redundant?

File: Sources/POI/OAPOIAdapter.h [hunk 43]
```diff
@@ -0,0 +1,40 @@
...
+//  OAPOIAdapter.h
+//  OsmAnd
+//
+//  Created by nnngrach on 11.08.2023.
+//  Copyright © 2023 OsmAnd. All rights reserved.
+//
+
+@class OAPOI;
+
+@interface OAPOIAdapter : NSObject
```
Comment (alex-osm) [inline Sources/POI/OAPOIAdapter.h:11]: Drop?

**Pull request merged**